src/utilities/HiFiMath.ts

```diff
@@ -0,0 +1,593 @@
+            z: f.b.x * g.a.z + f.b.y * g.b.z + f.b.z * g.c.z});
+        let c = new Vector3({
+            x: f.c.x * g.a.x + f.c.y * g.b.x + f.c.z * g.c.x,
+            y: f.c.x * g.a.y + f.c.y * g.b.y + f.c.z * g.c.y,
+            z: f.c.x * g.a.z + f.c.y * g.b.z + f.c.z * g.c.z});
+
+        return new Matrix3(a, b, c);
+    }
+
+    /**
+     * return the rotation part of the Quaternion in Matrix3 form
+     */
+    static fromQuaternion(q: Quaternion): Matrix3 {
+        // from https://www.euclideanspace.com/maths/geometry/rotations/conversions/quaternionToMatrix/hamourus.htm
+        let sqw = q.w * q.w; 
+        let sqx = q.x * q.x; 
+        let sqy = q.y * q.y; 
+        let sqz = q.z * q.z; 
+
+        // rotation matrix is scaled by quaternion length squared
+        // so we can avoid a sqrt operation by multiplying each matrix element by inverse square length
+        let invs = 1.0 / (sqx + sqy + sqz + sqw); 
+ 
+        let a = new Vector3();
+        let b = new Vector3();
+        let c = new Vector3();
+        a.x = ( sqx - sqy - sqz + sqw) * invs; 
+        b.y = (-sqx + sqy - sqz + sqw) * invs; 
+        c.z = (-sqx - sqy + sqz + sqw) * invs; 
+ 
+        let tmp1 = q.x * q.y; 
+        let tmp2 = q.z * q.w; 
+        b.x = 2.0 * (tmp1 + tmp2) * invs; 
+        a.y = 2.0 * (tmp1 - tmp2) * invs; 
+ 
+        tmp1 = q.x * q.z; 
+        tmp2 = q.y * q.w; 
+        c.x = 2.0 * (tmp1 - tmp2) * invs; 
+        a.z = 2.0 * (tmp1 + tmp2) * invs; 
+
+        tmp1 = q.y * q.z; 
+        tmp2 = q.x * q.w; 
+        c.y = 2.0 * (tmp1 + tmp2) * invs; 
+        b.z = 2.0 * (tmp1 - tmp2) * invs; 
+
+        return new Matrix3(a, b, c);
+    }
+
+    /**
+     * return the rotation part of the Matrix3 in Quaternion form
+     */
+    static toQuaternion(m: Matrix3): Quaternion {
+        // from https://www.euclideanspace.com/maths/geometry/rotations/conversions/matrixToQuaternion/
+        let q = new Quaternion();
+        q.w = Math.sqrt(Math.max(0.0, 1.0 + m.a.x + m.b.y + m.c.z )) / 2.0;
+        q.x = Math.sqrt(Math.max(0.0, 1.0 + m.a.x - m.b.y - m.c.z )) / 2.0;
+        q.y = Math.sqrt(Math.max(0.0, 1.0 - m.a.x + m.b.y - m.c.z )) / 2.0;
+        q.z = Math.sqrt(Math.max(0.0, 1.0 - m.a.x - m.b.y + m.c.z )) / 2.0;
+        if (m.c.y - m.b.z < 0.0) {
+            q.x *= -1.0;
+        }
+        if (m.a.z - m.c.x < 0.0) {
+            q.y *= -1.0;
+        }
+        if (m.b.x - m.a.y < 0.0) {
+            q.z *= -1.0;
+        }
+        return q;
+    }
+}
+
+// helper function that keeps an angle expressed in degrees in the range ]-360, 360[
+function sanitizeAngleDegrees(v: number): number {
+    // in the case v is Infinity or Nan,  let's special case
+    if (isNaN(v) || v === Infinity) {
+        return 0.0;
+    } else if (v === -Infinity) {
+        return -0.0;
+    } else {
+        // bring the value in the range ]-360, 360[
+        // if v is < 0 then it will cycle in ]-360, 0]
```

Comment:
Interestingly, -360 % 360 yields -0 according to my browser's javascript console (as does -720 % 360)

Although the difference is likely not meaningful, since -0 === 0 yields true.